core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -284,12 +289,18 @@ class KafkaController(val config : KafkaConfig, zkUtils: ZkUtils, val brokerStat
                   }
                   // If the broker is a follower, updates the isr in ZK and notifies the current leader
                   replicaStateMachine.handleStateChanges(Set(PartitionAndReplica(topicAndPartition.topic,
-                    topicAndPartition.partition, id)), OfflineReplica)
+                    topicAndPartition.partition, id)), OfflineReplica, newBatch = false)
                 }
               }
             }
           }
       }
+
+      inLock(controllerContext.controllerLock) {
```

Comment:
@lindong28 : Is the speedup really at step (a)? As we discussed earlier, to complete a ControlledShutdown request, the controller just needs to put 3000 requests into the ControllerChannelManager's in memory queue. It doesn't need to wait for those requests to be sent to the broker.

RequestSendThread currently only allows 1 outstanding request to a broker at a time.